### ISSUE TITLE
Plugin should update `contenthash` after minification

### ISSUE THREAD

**Issue by jakub-g** [webpack-contrib/terser-webpack-plugin#18]
This is a behavior that was probably in place forever with uglify hence it remains with terser as well.

### versions

* webpack Version: 4.20.1
* terser-webpack-plugin Version: 1.1.0

### config

```diff
    output: {
        filename = 'myprefix.[name].[contenthash].js',
        chunkFilename = 'myprefix.[name].[contenthash].js'
    },
    optimization: {
-       minimize: false,
+       minimize: true,
        minimizer: [new TerserPlugin()],
    }
```

### Expected Behavior

- When minification is on, `contenthash` of the output files should be different compared to when it is off

### Actual Behavior

- Regardless if minification is on or off, same `contenthash` is generated. i.e. file names are the same

This is problematic in case of scenario of upgrading the minificator.
- Since minification has no impact on the final file name, minifying with plugin version N vs plugin version N+1 might yield different content out of the same input file, yet same file name.
- Problematic in scope of long term caching, particularly in case of a bug in minificator. This can yield weird problems in deployment pipelines, CDNs etc. or lead to situation where different users have file with same name but different contents, depending on when they access the files and how the caching headers are configured.

To sum up, hard to diagnose bugs can happen if you're not aware of this issue.

According to https://github.com/webpack/webpack/issues/7483#issuecomment-394744914

>  plugin should regenerate `contenthash` if they change content of file

**Comment by evilebottnawi** [comment]
@jakub-g sounds intersting, can you provide minimum reproducible repo where contenthash is same for difference content, maybe bug in webpack

**Comment by michael-ciniawsky** [comment]
```
+ minimize: true
```

Don't do that as this will add the `uglifyjs-webpack-plugin` (default) to `plugins` behind the scenes (`webpack =< v4.0.0`), it's either `minimize` **or** `minimizer: [ ...plugins ]` 

`terser-webpack-plugin` reuses the (original) asset `filename`, so content changes aren't reflected in the `hash` portion of the `filename` at all atm. Not sure if tapping into the related `contentHash` hooks provided by `webpack` would work as a solution here (but maybe)

**Comment by dreyks** [comment]
@michael-ciniawsky yes plugin should tap into `compilation.hooks.contentHash.tap` to update the hash

**Comment by michael-ciniawsky** [comment]
Did you try this? According to https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L319 it get the `chunk` passed as an argument which is should already be rendered when the `terser-webpack-plugin` is applied

**Comment by evilebottnawi** [comment]
@dreyks friendly ping

**Comment by dreyks** [comment]
ugh... no spare time currently sorry

**Comment by edmorley** [comment]
In addition to the use-case above of upgrading minifier version to fix a minification bug, this issue also affects changing minification options. eg: adjust `terserOptions` and the generated `contenthash` stays the same.

**Comment by evilebottnawi** [comment]
Right now, it is impossible change `contenthash` based on `minified` output (because when we optimize files, we already have assets like `my.1a2b3c.js` where `1a2b3c` is content hash), i think we can fix it in `webpack@5` (need internal refactor). But we can update hash based on `minified` version and ``minified` options, ti should solve problem partially

**Comment by evilebottnawi** [comment]
Partial fixed https://github.com/webpack-contrib/terser-webpack-plugin/pull/44, let's leave the problem open because we need regerate contenthash based on minified output, it is require some change in `webpack`. Now we regenerate `contenthash` when `terser` was updated or `terserOptions` was changed.

**Comment by jakub-g** [comment]
@evilebottnawi thanks for picking this, I updated the plugin to 1.2.1 and indeed it solves the cases you mentioned.

Should I open a linked ticket on a webpack repo for the refactor you mentioned?

**Comment by evilebottnawi** [comment]
@jakub-g it is already in todo, maybe we solve this problem for `webpack@5` :+1:

**Comment by evilebottnawi** [comment]
Want to clarify, here 2 problems:
- We don't update `contenthash` after build (SOLVED)
- Calculate `contenthash` after file generation (should be solved on webpack size)
